### PR TITLE
Notifications

### PR DESCRIPTION
**Pull request by patriknordlen**
This PR addresses issue #228.

It adds a new model "notifications" which has a number of events that can be set up to trigger notifications to be sent by mail and/or as a Slack message. This can be done globally as well as on a per-user basis, so individual users can set up notifications of their own without needing to request assistance from an admin. Destinations for personal notifications are the user's mail address and Slack username, respectively.

Initial events that have been setup to offer notifications are:
- Engagement added
- Test added
- Results added to test
- Report created

Slack messages are sent using "django_slack" which has been added to the INSTALLED_APPS section of settings.dist.py.

### CODE REVIEW

**Comment by devGregA** [comment]
@grendel513 do you have time to review?

**Comment by devGregA** [comment]
Hi @patriknordlen awesome stuff as usual :) would you please add a section to https://github.com/OWASP/django-DefectDojo/blob/master/docs/features.rst documenting how to setup the integration?

https://github.com/OWASP/django-DefectDojo/blob/master/docs/features.rst generates
http://defectdojo.readthedocs.io/en/latest/features.html

**Comment by grendel513** [comment]
Does this replace or integrate with the current alerts in dojo?

I think it should replace them, notifications and alerts seem redundant and
alerts needs a revamp anyway.

**Comment by patriknordlen** [comment]
@grendel513 it does not replace alerts, it's a pure addition only. It wouldn't be too hard to add alert as a notification type in addition to the ones already in there...I haven't looked into revamping alerts though and build something that automatically does something the other way around (create notifications for anything that pushes an alert), that could possibly be more work. :)

**Comment by grendel513** [comment]
I'm not too keen on having two separate alert/notification mechanisms in
dojo...we need to work on combining them and making sure both
functionalities remain.  Since your notification mechanism appears to be
more mature/robust, I recommend taking a look at the current alert
functionality and finding a way to combine them.

I think this will confuse users who are not aware they are separate an
unrelated.

my two cents

**Comment by patriknordlen** [comment]
Definitely sounds fair. I don't mind working something out, just want to make sure I go for an approach that seems sensible to others as well.

My first thought would be my suggestion above, that is, have alerts work as another notification action. So users could then set up either mail, slack or alert as notification for events. Things that use alerts directly today would then need to be rewritten to use notifications instead, and the logic for which action to take would then be handled there.

**Comment by grendel513** [comment]
I think that sounds like an appropriate approach.  See
https://github.com/OWASP/django-DefectDojo/issues/220 for additional
comments on alerts.  You can see what type of events/situations trigger an
alert in the code here
https://github.com/OWASP/django-DefectDojo/blob/master/dojo/utils.py#L612

I feel we should incorporate those events as well.

Thanks for your willingness to take this on!! We really appreciate it.

Jay

**Comment by devGregA** [comment]
I agree that in the long run they should be one entity, but I personally would like slack integration sooner rather than later ¯\_(ツ)_/¯. @grendel513 would you be okay with merging as if functional and then we can create a separate issue?

**Comment by devGregA** [comment]
I also think before we combine them we would need to decide what displays where. Right now alerts serve to notify you when 1. something fails 2. something hasn't been completed. As an admin I wouldn't want notifications every time something is opened / closed etc. I see a lot of granularity that needs to be defined that is outside of the issue @patriknordlen is working to solve.

**Comment by grendel513** [comment]
the key phrase here is "in the long run" - it is my experience that those
things left undone become technical debt...if we are in there now, we
should get it done ;)

**Comment by patriknordlen** [comment]
One thing about current alerts is that they're not necessarily triggered by events. Things like stale or upcoming engagements/scans are not "generated" so much as they're returned when specifically asking for that info. To move that over to run via notifications there would have to be something like a cron job that runs regularly and collects the same info instead. Unless there's some nifty Django feature that I don't know about that solves this.

**Comment by devGregA** [comment]
@aaronweaver would you please share your thoughts?

**Comment by patriknordlen** [comment]
I've reworked notifications and alerts a bit so they're integrated now. As for notifications that would require some kind of period check - I'm thinking celery periodic tasks might be possible to use for that? My knowledge of it is limited to that it came up when I googled for if that kind of thing is possible with celery. Opinions, thoughts?

**Comment by grendel513** [comment]
Yes, celery is the right approach, and it is already a part of dojo for
reports and jira.  That should be the way to create tasks that are executed
at regular intervals -
http://docs.celeryproject.org/en/latest/userguide/periodic-tasks.html

**Comment by patriknordlen** [comment]
Thanks for the input @grendel513, I'll get right on it. :)

**Comment by grendel513** [comment]
take a look at
https://github.com/OWASP/django-DefectDojo/blob/master/dojo/tasks.py to see
how these are implemented.  additional tasks can be added to this file to
take care of the notifications.  thanks again for attempting to get this
done

**Comment by patriknordlen** [comment]
Just a short update on this. I've worked on it a bit and have some local changes that I haven't committed yet because I'm still rewriting things here and there.

- I've added alerts as another notification option.
- I've changed all parts of the application that notify the user by mail or alert in some way to use the new notification functionality instead.
- I've implemented periodic tasks with celery that will check things that were previously in get_alerts and send notifications based on that.

In other words, progress is being made, so no need to merge this in its current state :)

**Comment by grendel513** [comment]
@patriknordlen just wondering how this was coming along?

**Comment by patriknordlen** [comment]
@grendel513 making progress now - I'm back from a 4w vacation which is why there's been a lack of updates. Functionality is done, working now on testing functionality in various situations and ensuring a smooth migration.

**Comment by patriknordlen** [comment]
Ok! Summary of the changes:

- All types of notifications in the application is now handled via the function 'create_notification' in utils.py. This function will in turn send notifications via mail, Slack and/or alerts if this has been enabled either on a global or personal scope in the Notifications page.
- Notifications are based on templates found in the dojo/notifications directory. Templates are matched using the 'event' argument to create_notification. If no template is found, the template 'other.tpl'  will be used.
- Alerts are no longer loaded in full when loading defectDojo pages. An alert count endpoint is called every 10 seconds, but full alert data is only loaded when the user clicks on the dropdown.
- The alerts page now has functionality to select and remove alerts.
- Alerts that were previously dynamically retrieved each time a user requested his/her alerts (upcoming engagements, stale engagements) have been moved to be run as periodic tasks using celery beat instead. These tasks will run every hour and create alerts according to their configuration.

**Comment by patriknordlen** [comment]
@devGregA @grendel513 please have a look - with a PR of this size I'm sure there are problems that I've missed even after testing. For instance I don't have a jira instance handy to see if everything works properly there.

**Comment by devGregA** [comment]
@aaronweaver would you also have time to test?

**Comment by grendel513** [comment]
@devGregA, @aaronweaver, @patriknordlen - I tested the UI as well as the Alerts and email integrations.  All work great.  I have not tested integration with Slack, Hipchat, or Jira.

I think we need to add documentation for this new feature...especially the need to:
    `pip install django-slack`
    `pip install django-multiselectfield`
    `./manage.py makemigrations`
    `./manage.py migrate`

We also need to make sure we mention that the celery process needs to be running for all of this to work.

Thanks!

**Comment by devGregA** [comment]
@grendel513 I have tested the JIRA changes, but not the notifications + JIRA. My suggestion is to get the documentation in as you noted, but then I would be okay with merging and looking at it as 'beta features' I have some code I want to submit relating to handling false positives, but I want to get this done first so @patriknordlen doesn't have to deal with conflicts. Would you be okay with merging once the docs are in place and treating them as 'beta' features for maybe a week? By then I will have plenty of time to thoroughly test the changes at my work.

**Comment by patriknordlen** [comment]
Sounds good @devGregA! I'll work on the docs tonight.

**Comment by patriknordlen** [comment]
@devGregA @grendel513 sorry for the delay - work etc. I've added a new section to the 'Features' page covering notifications in brief detail.

`django_slack` is now gone as a dependency (it's not used anymore, notifications are now sent directly using the `requests` library) and django-multiselectfield is added as a dependency in `setup.py`. Users are fine if they run `pip install .` which will take care of installing all dependencies for them, so I added that command to the list of default commands to run when upgrading. `makemigrations` and `migrate` are already there.

**Comment by devGregA** [comment]
Merged. Will roll out at my employer this week for testing.
